### ISSUE TITLE
Don't use actions-anti-pattern for filter box rendering

### ISSUE THREAD

**Issue by jrieken** [microsoft/vscode#162304]
continues https://github.com/microsoft/vscode/issues/162004#issuecomment-1261139593

There seems to be a anti-pattern that noop actions are registered to views for the sole reason of rendering a custom UI for a single place. This is bad because

* actions are global and discoverable, e.g users can register keybinding to them or extensions can invoke them
* action have certain semantics and new features, like hiding, don't make sense
* this seems like a lot of complexity and overhead to just render something into the panel title

As tactical fix for #162004 I have marked some actions as fake action. This is debt and needs to be replaced with a proper solution. I suggest to either

* fix the actions to do something, e.g make them work independent of the one special rendering
* don't use an action but let your view override [renderHeader](https://github.com/microsoft/vscode/blob/2c801254ebf547baaee771b99aa76d202f999622/src/vs/workbench/browser/parts/views/viewPane.ts#L287) and add custome UI, e.g a filter box

---

- [ ] src/vs/workbench/contrib/comments/browser/commentsViewActions.ts @alexr00 
- [ ] src/vs/workbench/contrib/debug/browser/repl.ts @roblourens @connor4312 
- [ ] src/vs/workbench/contrib/markers/browser/markers.contribution.ts @lszomoru 
- [ ] src/vs/workbench/contrib/output/browser/output.contribution.ts @sandy081 
- [ ] src/vs/workbench/contrib/testing/browser/testingExplorerFilter.ts @connor4312

**Comment by sandy081** [comment]
I do not think `renderHeader` works because the panel view header and the panel toolbar are not the same.

**Comment by sandy081** [comment]
@jrieken I think either of the options mentioned above are helpful/feasible. From beginning, panels do not have access to panel title bar to render the widgets. So we would need to introduce back `getActions` api for views from which we moved away in favour of menus. I also think some of these actions like output switching are commands with args that means they can support registering keybindings. 

Is it possible to have a new concept in the menu registry to accommodate those actions which needs special rendering? Such actions can also support hiding but are not shown in the overflow menu?

**Comment by jrieken** [comment]
> Such actions can also support hiding but are not shown in the overflow menu?

I don't believe that is a good solution. You are not a valid action you require special rendering and cannot work else. Let's not forget that any action can always be invoked from anyone form anywhere. 

> I also think some of these actions like output switching are commands with args that means they can support registering keybindings.

It is great if argument-less invocation can be supported and I totally see how switching output is a perfect case for that - in the lack of an argument show quick pick with all options but for filtering I don't see a good way. 

> From beginning, panels do not have access to panel title bar to render the widgets. So we would need to introduce back getActions api for views

Why must this be an action? If `renderHeader` isn't a good fit can we add something fit for the purpose? Except for output switching (see above) this is always about filtering. Why isn't there a get filter-ui method that can be overridden?

**Comment by Tyriar** [comment]
@sandy081 `SwitchTerminal` is this when tabs are disabled:

![image](https://user-images.githubusercontent.com/2193314/193599325-e758d3c4-8d7b-4d5e-a9ff-7799cdd19cf5.png)

I'll watch this issue to see if the fix you end up doing for this can be applied there too.

**Comment by sandy081** [comment]
Adopted following actions

### Select Action

Changed the output select action to a submenu with `selection` flag. Contributed the selection entries to this submenu. Implemented a new `SubmenuEntrySelectActionViewItem` that renders this select submenu as selection dropdown. This will enable user hiding this selection action and will get moved into the secondary actions as below

![image](https://user-images.githubusercontent.com/10746682/194834109-44ccfbda-6d3f-4513-ae93-037a56eaea88.png)

Following this example for reference: 

https://github.com/microsoft/vscode/blob/150b7f9d0ba9d3cdd294a47b164bd22fb5017ed2/src/vs/workbench/contrib/output/browser/output.contribution.ts#L116-170

@roblourens @Tyriar Please follow above example to convert your select actions.

### FilterViewPane

Moved filter into ViewPane and called it `FilterViewPane` that supports rendering the filter widget in the header/body. Views supporting filtering can extend this. Eg: Problems, Comments & Repl. Adopted it in these views. CC @alexr00 @roblourens

**Comment by roblourens** [comment]
Thanks for adopting this for me! But it's still basically based on a "fake action" right? https://github.com/microsoft/vscode/pull/163132/files#diff-3e738fc4417c6ddcfa32cbda3356cb2b2b30943137d429757e3d7585764f22b9R60

**Comment by sandy081** [comment]
That's right. It is based on old action and an implementation detail of the view and not leaking into menus.

**Comment by alexr00** [comment]
Thanks for adopting this for me too!

**Comment by jrieken** [comment]
>  But it's still basically based on a "fake action" right?

I'd say we have two "kind of actions". There is `Action2` which should be seen as a model object, they are singletons, can be placed in menus, and triggered by keybindings. Then there is `IAction` which is view model object which is used for rendering only (well, mostly since its definition has been a bit fluent). A sample is Separator which is an IAction but shouldn't ever be a "model object action"

**Comment by roblourens** [comment]
I think this is complete
